Release 0.2

jvatic opened the pull request:
@titanous

Commit:
Prevent duplicate mentions for a single post

fixes #43

Commit:
Proxy attachment requests for non original posts when following
associated with post

Commit:
Add Count header

closes tent/tent.io#134

Commit:
POST /followers refreshes existing relationship

closes tent/tent.io#122

Commit:
Add /followers/:entity 302 redirect to /followers/:id

refs tent/tent.io#143

Commit:
Add GET /followings/:entity redirect to /followings/:id

closes tent/tent.io#143

Commit:
Add COUNT header to HEAD /posts/:post_id/versions

refs tent/tent.io#134

Commit:
Add pagination in link header for GET/HEAD /posts

Commit:
Add pagination in link header for GET/HEAD /followings

Commit:
Add pagination in link header for GET/HEAD /followers

Commit:
Add pagination in link header for GET/HEAD /groups

Commit:
Add GET /posts/:post_id/mentions endpoint

- only returns mentions referencing a public post known to server
- supports pagination via
  before_id/before_id_entity/since_id/since_id_entity/limit params
- sets COUNT header for HEAD requests
- sets pagination in link header

Commit:
Standardize JSON errors

{ "error": "{explanation}" }

Commit:
Add optional 'tent_expires_at' to app authorizations and issue a
'refresh_token' along with mac credentials

Commit:
Add profile info versions

Commit:
Add GET /profile/:type_uri[?version=n]

Commit:
Fix: set Content-Type header for error responses

Commit:
Fix: return 403 for expired app authorizations

Commit:
Fix app authorization mac_key_id prefix

Commit:
Update spec

Commit:
Fix GET /followings/:entity redirect

Commit:
Fix GET /followers/:entity redirect

Commit:
Update profile info versions migration to create first version for
existing profile infos

Commit:
Fix API::Middleware#self_uri

Commit:
Require models instead of autoload (fixes weirdness)

Commit:
Fix triggering notifications

fixes #49

Commit:
Notify followings of profile updates

refs #48

Commit:
Fix: send deleted post notifications to mentioned entities

Commit:
Fix: don't notify mentions of deleted post until permissions set

Commit:
Fix entity propagation for followers and followings

Commit:
Fix Followings::Discover and ::Follow error response format

Commit:
Fix user scoping for POST /followings

Commit:
Fix user scoping

titanous commented on lib/tentd/api/authentication_lookup.rb:
```diff
@@ -18,14 +18,14 @@ def action(env)
           env.hmac.algorithm = env.potential_auth.mac_algorithm
         elsif mac_key_id
-          return [401, {'WWW-Authenticate' => 'MAC'}, ['Invalid MAC Key ID']]
+          return [401, {'WWW-Authenticate' => 'MAC', 'Content-Type' => MEDIA_TYPE}, [{ 'error' => 'Invalid MAC Key ID' }.to_json]]
```
How about a helper function that handles error responses?

titanous commented on lib/tentd/api/followers.rb:
```diff
@@ -3,8 +3,26 @@ class API
+        def action(env)
+          id_or_entity = env.params.delete(:captures).first
+
+          if id_or_entity =~ /^https?:\/\//
```
Prefer `\A` over `^`.

titanous commented on lib/tentd/api/followers.rb:
```diff
@@ -110,8 +129,31 @@ def action(env)
+          follower = Model::Follower.select(:id, :public, :public_id).where(:entity => env.params.follower_entity, :user_id => Model::User.current.id).first
+          if follower && !follower.public? && !(env.full_read_authorized || authorize_env?(env, :self))
+            follower = Model::Follower.find_with_permissions(follower.id, env.current_auth)
+            raise Unauthorized unless follower
```
This should actually 404.

titanous commented on lib/tentd/api/followers.rb:
```diff
@@ -110,8 +129,31 @@ def action(env)
+          end
+
+          unless follower
+            raise Unauthorized unless env.full_read_authorized
```
I think this should just 404 as well.

jvatic replied:
Then [`GetOne`](https://github.com/tent/tentd/blob/9aee6cfca3ae743efcf5e000b31ad6aee8fa2bd5/lib/tentd/api/followers.rb#L166) should 404 also?

titanous replied:
Yes.

titanous commented on lib/tentd/api/followers.rb:
```diff
@@ -147,6 +189,22 @@ def action(env)
+        def action(env)
+          count_env = env.dup
+          count_env.params.return_count = true
+          count = GetMany.new(@app).call(count_env)[2][0]
```
This is weird. Also, this header should only be added on HEAD requests.

titanous replied:
Missed that this happens only on head requests. Why does `GetMany` get called twice?

jvatic replied:
The first call to `GetMany` is the same as a GET request (sets the Content-Length header), and the second call gets the count (which is not the same as `env.response.size`.)

titanous replied:
Ah.

titanous commented on lib/tentd/api/followings.rb:
```diff
@@ -26,6 +44,27 @@ def action(env)
+          end
+
+          unless following
+            raise Unauthorized unless authorize_env?(env, :read_followings)
```
Ditto on the 404s here.

titanous commented on lib/tentd/api/middleware.rb:
```diff
@@ -25,7 +29,24 @@ def call(env)
+      end
+
+      def serialize_params(params)
+        "#{params.inject([]) { |m, (k,v)| m << "#{k}=#{URI.encode_www_form_component(v)}"; m }.join('&')}"
```
Why is this wrapped in `"#{}"`?

titanous commented on lib/tentd/api/pagination_header.rb:
```diff
@@ -0,0 +1,53 @@
+        env['response.headers'] ||= {}
+        pagination = [%(<#{prev_uri(env).to_s}>; rel="prev"), %(<#{next_uri(env).to_s}>; rel="next")]
+        if env['response.headers']['Link']
+          env['response.headers']['Link'] += ",#{pagination.join(',')}"
```
Add a space after the first `,`.

titanous commented on lib/tentd/api/pagination_header.rb:
```diff
@@ -0,0 +1,53 @@
+        if env['response.headers']['Link']
+          env['response.headers']['Link'] += ",#{pagination.join(',')}"
+        else
+          env['response.headers']['Link'] = pagination.join(',')
```
`.join(', ')` here and above.

titanous commented on lib/tentd/api/pagination_header.rb:
```diff
@@ -0,0 +1,53 @@
+        params = env.params.dup
+        resource = env.response.last
+
+        params.before_id = resource.public_id
```
Check if we're on the last page by checking whether the size of the response is the size of the requested page size.

jvatic replied:
I also missed the case when `before_id` and `since_id` have reversed roles when ordering `asc` vs `desc`.

Commit:
Cleanup API::Middleware#serialize_params

Commit:
Add API::Middleware#error_response helper

titanous commented on lib/tentd/api/middleware.rb:
```diff
@@ -25,7 +29,28 @@ def call(env)
+        params.inject([]) { |m, (k,v)| m << "#{k}=#{URI.encode_www_form_component(v)}"; m }.join('&')
+      end
+
+      def error_response(status, headers, error)
```
Perhaps you could put the headers last and default so that they can be omitted unless needed?

jvatic replied:
Yes, I was trying to keep with the status, headers, body convention but I suppose that doesn't really matter in this case.

titanous commented on lib/tentd/api/posts.rb:
```diff
@@ -277,24 +325,87 @@ def action(env)
+        end
+      end
+
+      class MentionsCountHeader < Middleware
```
It seems like these count header middlewares have a lot of boilerplate, is it possible to DRY it up a bit?

Commit:
Add spaces to pagination Link header after ','s

Commit:
Refactor API::Middleware#error_response helper

Commit:
Use \A instead of ^

Commit:
Fix: /followers/:id_or_entity returns 404 when not found instead of 403

Commit:
Fix: /followings/:id_or_entity returns 404 when not found instead of 403

Commit:
Refactor count header middleware

Commit:
Refactor Posts::MentionsPaginationHeader

Commit:
Lock fabrication gem at v2.3.0

titanous commented on db/migrations/003_add_tent_expires_at_to_app_authorizations.rb:
```diff
@@ -0,0 +1,8 @@
+Sequel.migration do
+  change do
+    alter_table(:app_authorizations) do
+      add_column :tent_expires_at, Integer
```
Can we just call this column `expires_at`?

Commit:
Rename app_authorizations tent_expires_at column to expires_at

Commit:
Refactor pagination header tests

Commit:
Fix pagination header for GET /posts when order=asc

Commit:
Fix pagination for GET /posts (map id to sort column internally for
before_id and since_id params)

Commit:
Fix GET /posts/:post_id/mentions pagination: order by mentioned post id

Commit:
Fix GET/HEAD /posts/:id/versions

- sort by version
- add since_version/before_version for pagination
- add pagination in link header

Commit:
Add created_at timestamp to follower json

Commit:
Add created_at to following json

Commit:
Add created_at to group json

Commit:
Add created_at to app json

Commit:
Don't inititalize db immediately on `require 'tentd'`

Commit:
Add tent_version to profile

Commit:
Fix entity redirects

Commit:
Update /followings/:entity and /followers/:entity to return 200 with
Content-Location header set instead of 302

Commit:
Fix following proxy

Commit:
Fix following proxy: use all servers and relative path

Commit:
Fix /posts/:id/mentions with post_types param
